core/src/Cuda/Kokkos_Cuda_Parallel.hpp

```diff
@@ -885,9 +885,14 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
   using value_type     = typename ValueTraits::value_type;
   using reference_type = typename ValueTraits::reference_type;
   using functor_type   = FunctorType;
-  using size_type      = Kokkos::Cuda::size_type;
-  using index_type     = typename Policy::index_type;
-  using reducer_type   = ReducerType;
+  // Conditionally set size_type to int16_t or int8_t if value_type is less than
+  // int32_t (Kokkos::Cuda::size_type)
+  using size_type = typename std::conditional<
```

Comment:
Good question. The name `size_type` was confusing to me as well. Here, `size_type` is <= the size of `value_type`. `size_type` is used to determine the word count, shared memory buffer size, and global memory buffer size before the reduction is performed. Within the reduction, atomics (other than implicit atomic stores and loads?) are only used for keeping track of metadata in `m_scratch_flags` which is why that local was left as type `Cuda::size_type*`. Also within the reduction, the word count is recomputed based on `size_type` and when calculating indexes into the shared/global memory buffers for performing the reduction, `size_type` is used again. So, without this compile-time conditional for setting `size_type`, our word_count and/or indexes are not correct for scalars < 4 bytes in size. For scalars > 4 bytes in size, indexing into shared/global memory relies on the block and grid dimensions to ensure that we index at the correct offset rather than at every 4 byte word; such that, when the join is performed, we have the correct data that was copied over in chunks of 4 bytes.

Christian suggested that I try out the approach you've used in #4077 for handling volatile reductions with `half_t` so I'll flag you in that PR after this merges.

Reply:
@PhilMiller: Does`word_size_type` make more sense?

Reply:
Yeah, `word_size_type` was the replacement I had in mind too.

I would support making that replacement in this PR, but support from @crtrott would probably mean more

Reply:
@crtrott: All the PR feedback changes requested from here and #4101 are sitting in my local branch. Do you want these pushed here or in a follow on?